client/executor/src/wasm_runtime.rs

```diff
@@ -216,7 +223,7 @@ impl RuntimeCache {
 			let instance_pool = runtime.instances.read().clone();
```

Comment:
I think we should change `instances` to 
```
instances: Vec<RwLock<Option<Arc<VersionedRuntime>>>
```

When creating the `VersionedRuntime`, we do the following:
```
let mut instances = Vec::with_capacity(max_instances);
instances.resize(max_instances, Default::default());
```

What is the advantage of this solution?:
- `max_size` field isn't required anymore
- We don't need to clone here the instances anymore. The code would just be:
```
let instance = runtime.instances.iter().
```
- We need to lock on mutex less.

Reply:
Unfortunately this will require `RwLock` to be locked for the duration of the call. See https://github.com/paritytech/substrate/pull/5177#discussion_r389156381

Reply:
Unfortunately mutex is still needed because `Arc<T>` will only implement `Send` and `Sync` as long as the `T` implements `Send` and `Sync`.

Additionally, `RwLock` cannot be cloned, so `resize` won't work.

Reply:
Shouldn't a Box for wasminstance be enough? And also make sure the that wasminstance implements Send? 

Than just use resize_with...

Reply:
`WasmInstance` implements `Send` but not `Sync` indeed, so you can't concurrently access a collection of them.

Reply:
I have implemented it, please take a look.